plugins/modules/storage/zfs/zfs.py

```diff
@@ -184,25 +184,33 @@ def set_property(self, prop, value):
             return
         cmd = [self.zfs_cmd, 'set', prop + '=' + str(value), self.name]
         (rc, out, err) = self.module.run_command(cmd)
-        if rc == 0:
-            self.changed = True
-        else:
+        if rc != 0:
             self.module.fail_json(msg=err)
 
     def set_properties_if_changed(self):
         current_properties = self.get_current_properties()
         for prop, value in self.properties.items():
             if current_properties.get(prop, None) != value:
                 self.set_property(prop, value)
+        if self.module.check_mode:
+            return
+        updated_properties = self.get_current_properties()
+        for prop in self.properties:
+            value = updated_properties.get(prop, None)
+            if value is None:
+                self.module.fail_json(msg="zfsprop was not present after being successfully set: " + prop)
+            if current_properties.get(prop, None) != value:
+                self.changed = True
 
     def get_current_properties(self):
-        cmd = [self.zfs_cmd, 'get', '-H']
+        cmd = [self.zfs_cmd, 'get', '-H', '-p', '-o', "property,value,source"]
         if self.enhanced_sharing:
             cmd += ['-e']
         cmd += ['all', self.name]
         rc, out, err = self.module.run_command(" ".join(cmd))
         properties = dict()
-        for prop, value, source in [l.split('\t')[1:4] for l in out.splitlines()]:
+        for line in out.splitlines():
+            prop, value, source = line.split('\t')
```

Comment:
i believe we could skip '\t' as a separator here and use just `split()` ...

Reply:
I think using is probably better. ZFS allows datasets to have spaces in their name, but literal tabs are not allowed. `zfs create 'rpool/foo bar'` will work, while `zfs create "rpool/foo\tbar"` will fail due to an invalid character.

For reference, the `zfs-get(8)` man page says this about the `-H` option:
```
Display output in a form more easily parsed by scripts.  Any headers are omitted, and fields are explicitly separated by a single tab instead of an arbitrary amount of space.
```

Reply:
omg, i didn't knew that zfs allowed spaces in volume names! yes, indeed, tab would be better (even we used `-o property,value,source`, `mountpoint` value would be with space as well, if volume name included space). thanks!

Reply:
I was surprised to discover that too! I think it's something that most reasonable users would never try, haha.